### ISSUE TITLE
Any differences between headless and non-headless mode?

### ISSUE THREAD

**Issue by mhp-borisbojic** [getgauge/taiko#1480]
### Expected behavior
An e2e test behaves the same in headless or observe mode

### Actual behavior
An e2e results in a different behavior in headless mode

### Versions
```
node v12.18.3
all taiko versions (including the latest)
```

### Description

Not really an issue, but more a question - does taiko do **anything** differently in the headless mode then in the observe mode (besides not opening a visible chrome window)? Other user-agent? Language specific things? Cookie / Session / Storage / Anything? :(

I'm asking because in our current e2e test, we receive different results when using the headless mode. Using the observe mode everything works as expected and defined.

### taiko configuration
```
const taikoTest = async (testName, test) => {
    console.log(testName);
    try {
        await t.openBrowser({
            args: [
                '--window-size=1920,1080',
                '--disable-gpu',
                '--disable-dev-shm-usage',
                '--disable-setuid-sandbox',
                '--no-first-run',
                '--no-sandbox',
                '--no-zygote'
                ]
        });
        await t.setConfig({
            navigationTimeout: 120000,
            observeTime: 1,
            retryTimeout: 30000
        });
        await test();
    } catch (e) {
        console.error(e);
        fs.mkdirSync('tmp/screenshots', {recursive: true});
        await screenshot({path: `tmp/screenshots/${timeNow}-${toFileName(testName)}.png`, fullPage: true});
        await closeBrowser();
        process.exit(1);
    } finally {
        await closeBrowser();
    }
};
```

**Comment by NivedhaSenthil** [comment]
Yes, not Taiko but the browser behaves differently in headless and headful modes

**Comment by mhp-borisbojic** [comment]
... and how? Other e2e tools that uses Chromium doesn't appear to have the issues we are facing at the moment?

**Comment by NivedhaSenthil** [comment]
That shouldn't be the case though, can you please give a sample and explain what exactly is happening differently ? To check if Taiko is doing anything differently.

**Comment by mhp-borisbojic** [comment]
We found out, more or less, how to fix our issue (but not the cause of it).

We changed our taiko config to:
```
setConfig({
            navigationTimeout: 120000,
            observe: true,
            observeTime: 3000,
            retryTimeout: 5000
        });
```

With this, we don't have the issues anymore.

**Comment by NivedhaSenthil** [comment]
This seems like a problem with waits since element on action is highlighted in headful mode it can take a little more time compared to headless mode, arbitrarily adding observeTime can heavily increase the test execution time. Please try waiting for proper events or states before an action, https://docs.taiko.dev/api/waitfor/

If you can provide us a sample we could try covering those cases in Taiko if possible or suggest a better way to wait instead of adding wait time between all actions by adding observe time.

**Comment by mhp-borisbojic** [comment]
The example is again the same login form ;)

Please try out this form: https://login.porsche.com/login/gb/en_GB#

In headless mode (without using observe: true in the config), logging in in does not work because (for whatever reason) after the password was entered.

Using the observe: true config, it works as expected.

The code is the same as in https://github.com/getgauge/taiko/issues/1476#issuecomment-690096001, but here some more:

```
const loginWithoutNavigation = async (url, username, password) => {

  await goto(url);
  await write(username, into(protractor('email-input')))
  await clickPassBtn();

  await write(password, into(protractor('password-input')))
  await clickLoginBtn();

  success('Login successful');
}
```

**Comment by NivedhaSenthil** [comment]
I could see that the page has lots of shadow dom contents to be loaded which takes time to get to right place, in headless mode since the execution is faster and elements are not in desired place I could see the actions are happening incorrectly. 
When I waited for `firstMeaningfulPaint` which ensures page's primary content appeared on the screen as per heuristics [here](https://docs.google.com/document/d/1BR94tJdZLsin5poeet0XoTW60M0SjvOJQttKT-JK8HI/view), I could see the script passing fine consistently. Below is the script I tried,
```
(async () => {
    try {
        await openBrowser();
        await goto("https://login.porsche.com/login/gb/en_GB",{waitForEvents:['firstMeaningfulPaint']});
        await write("xxx", into(textBox(below("Porsche ID (email address)"))));
        await click("Continue and enter password");
        await write("xxx",into(textBox(below("Password"))));
        await click("Log in now");
        await text("user name").exists();
    } catch (error) {
        console.error(error);
    } finally {
        await closeBrowser();
    }
})();
```
PS: Taiko uses `loadEventFired` as default event to wait for which ensures page load. And there were issues where `firstMeaningfulPaint` event was [not triggered by the browser](https://github.com/getgauge/taiko/issues/393) hence moved from using that as default letting user choose to use it as per need.
